blog/_posts/2024-08-23-fine-tuning-language-models-with-kaito.md

```diff
@@ -0,0 +1,145 @@
+
+Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.
+
+## KAITO can help with model fine-tuning
+
+KAITO supports Low-Rank Adaption (LoRA) and Quantized LoRA (QLoRA), the [most common](https://arxiv.org/pdf/2402.16141) parameter-efficient fine-tuning techniques, which significantly lower the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 
+
+Let’s start the journey of fine-tuning the Phi-3 model using KAITO.
+
+First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/sidddddddddddd/kubernetes) that is optimized for knowledge about Kubernetes and cloud platforms.
+
+> This is the original dataset which we selected and reformatted using [these instructions](https://github.com/Azure/kaito/tree/main/docs/tuning#input-dataset-format) from the HuggingFace trainer library.
+
+To use the data set in KAITO, we can specify it in the new **tuning** property in the workspace API. 
+
+Here’s an example of what the fine-tuning workspace looks like when using a public dataset from HuggingFace:
+
+```bash
+apiVersion: kaito.sh/v1alpha1
+kind: Workspace
+metadata:
+  name: workspace-tuning-phi-3
+resource:
+  instanceType: Standard_NC24ads_A100_v4
+  labelSelector:
+    matchLabels:
+      app: tuning-phi-3
+tuning:
+  preset:
+    name: phi-3-mini-128k-instruct
+  method: qlora
+  input:
+    urls:
+      - https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true
+  output:
+    image: myregistry.azurecr.io/adapters/myadapter:0.0.1
+    imagePushSecret: myregistrysecret
+```
+
+The tuning property specifies the minimum information required for starting a tuning job, including:
+
+**Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 
+
+**Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 
+
+**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).
+
+**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.
+
+> The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 
```

Comment:
Is there ever an option to map LLM's to GPU's that they "fit into" for finetuning.  A/H100 are very expensive and potentially overkill for a Phi-3 for example

Reply:
This is not a blocker by the way, just a "wouldn't it be nice"

Reply:
Good question - for inferencing with Phi-3 just the NC6 series GPU is needed, but KAITO currently needs A100 or higher to fine-tune in a reasonable amount of time (3 hours or less), and in the future more optimizations will be made to lower this requirement.